File: Python/MST.py
import heapq

class Edge:
    def __init__(self,x,y,dst):
        self.x=x
        self.y=y
        self.dst=dst
    def __lt__(self,other):
        return self.dst<other.dst

def Find(x):
    if x==arr[x]:
        return x
    arr[x]=Find(arr[x])
    return arr[x]

def Union(x, y):
    arr[Find(x)]=Find(y)

def Kruskal():
    ret=0
    while pq:
        top=heapq.heappop(pq)
        x,y,curr=top.x,top.y,top.dst
        if Find(x)==Find(y):
            continue
        Union(x,y)
        ret+=curr
    return ret

N,M=map(int,input().split())
arr=[i for i in range(N+1)]
pq=[]
for _ in range(M):
    a,b,c=map(int,input().split())
    heapq.heappush(pq,Edge(a,b,c))
print(Kruskal())

File: Java/MST.java
import java.util.*;

class Edge implements Comparable<Edge>{
    int x, y, dst;
    Edge(int x, int y, int dst){
        this.x=x;
        this.y=y;
        this.dst=dst;
    }
    @Override
    public int compareTo(Edge other){
        return Long.compare(this.dst,other.dst);
    }
}

public class MST{
    static int N, M;
    static int[] arr=new int[1005];
    static PriorityQueue<Edge> pq=new PriorityQueue<>();

    static int Find(int x){
        if (x==arr[x])
            return x;
        return arr[x]=Find(arr[x]);
    }

    static void Union(int x, int y){
        arr[Find(x)]=Find(y);
    }

    static int Kruskal(){
        int ret=0;
        while (!pq.isEmpty()){
            Edge edge=pq.poll();
            int x=edge.x, y=edge.y, curr=edge.dst;
            if (Find(x)==Find(y))
                continue;
            Union(x,y);
            ret+=curr;
        }
        return ret;
    }

    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        N=sc.nextInt();
        M=sc.nextInt();
        for (int i=0;i<=N;i++)
            arr[i]=i;
        for (int i=0;i<M;i++){
            int a=sc.nextInt();
            int b=sc.nextInt();
            int c=sc.nextInt();
            pq.add(new Edge(a,b,c));
        }
        System.out.println(Kruskal());
        sc.close();
    }
}
